### PR TITLE
Support for uploadpack.allowReachableSHA1InWant introduced in git 2.5

### PR DESCRIPTION
**Pull request by joshburt**
Hello!  
We're introducing some deployments that leverage a new git feature added in version 2.5: uploadpack.allowReachableSHA1InWant, https://github.com/git/git/blob/v2.5.0/Documentation/config.txt#L2570

Summary
-----------------------------------------------------------
In order to accomplish this I've updated the git HWRP.  I thought you might enjoy the additional feature, bug fixes, and polish that was done along the way.

To ease review and management of the changes I've split all the work into several pull requests.  This first one adds a number of linting improvements to the current code base as well as uploadpack.allowReachableSHA1InWant support.

The remaining fixes are based off this PR and cover:
CHEF-3929 https://github.com/chef/chef/issues/3929
CHEF-4126 https://github.com/chef/chef/issues/4126
Once this PR is merged the other two can be submitted.

High Level Changes
----------------------------------------------------

1. Consolidated usage of shell_out
2. Consolidated git command building code
3. rubocop linting offenses are way down!
4. lots of methods moved to private that don't appear to have a need to be public
5. Unit tests updated
6. Added ability to fetch by any ref (if supported) introduced in git 2.5

Areas of Discussion
-----------------------------------------------------------

- Cooler attribute name than "uploadpack_allow_reachable_sha1_in_want"
- Linting issues encountered by your build/ci system (if any)
- whyrun support was not updated

I'd love some help with validation, feedback, etc.  Thanks!
Josh

### CODE REVIEW

File: lib/chef/provider/git.rb [hunk 1]
```diff
@@ -16,15 +17,15 @@
 # limitations under the License.
 #
 
-require "chef/exceptions"
-require "chef/log"
-require "chef/provider"
-require "fileutils"
+require 'chef/exceptions'
```
Comment (coderanger) [inline lib/chef/provider/git.rb]: Pretty sure this will be rejected by Chefstyle.

**Comment by coderanger** [comment]
Please don't mix major code changes with purely stylistic changes. It makes reviewing the diff a million times more annoying.

**Comment by coderanger** [comment]
Also I think you might be using stock RuboCop, which is not what we use. Check out Chefstyle instead.

**Comment by joshburt** [comment]
That was the feedback I needed.  I'll split this into 2 PRs; the first for the linting offenses, and second for the feature enhancement.  Closing this PR.  Thanks Noah! :)

**Pull request closed without merging**